src/SelfServe/SelfServeComponent.tsx

```diff
@@ -312,41 +322,52 @@ export class SelfServeComponent extends React.Component<SelfServeComponentProps,
-                  {this.state.refreshResult.notificationMessage}
-                </MessageBar>
-              )}
-              {this.state.notification && (
-                <MessageBar
-                  messageBarType={getMessageBarType(this.state.notification.type)}
-                  styles={{ root: { width: 400 } }}
-                  onDismiss={() => this.setState({ notification: undefined })}
-                >
-                  {this.state.notification.message}
-                </MessageBar>
-              )}
-              <SmartUiComponent
-                disabled={this.state.refreshResult?.isUpdateInProgress}
-                descriptor={this.state.root as SmartUiDescriptor}
-                currentValues={this.state.currentValues}
-                onInputChange={this.onInputChange}
-                onError={this.onError}
-              />
-            </>
-          )}
-        </Stack>
-      </div>
+      <Translation>
+        {(translate) => {
+          const getTranslation = (key: string): string => {
+            return translate(`${this.smartUiGeneratorClassName}.${key}`);
+          };
+
+          return (
+            <div style={{ overflowX: "auto" }}>
+              <Stack tokens={containerStackTokens} styles={{ root: { padding: 10 } }}>
+                <CommandBar styles={{ root: { paddingLeft: 0 } }} items={this.getCommandBarItems(translate)} />
+                {this.state.isInitializing ? (
+                  <Spinner
+                    size={SpinnerSize.large}
+                    styles={{ root: { textAlign: "center", justifyContent: "center", width: "100%", height: "100%" } }}
+                  />
+                ) : (
+                  <>
+                    {this.state.refreshResult?.isUpdateInProgress && (
+                      <MessageBar messageBarType={MessageBarType.info} styles={{ root: { width: 400 } }}>
+                        {getTranslation(this.state.refreshResult.notificationMessage)}
+                      </MessageBar>
+                    )}
+                    {this.state.notification && (
+                      <MessageBar
+                        messageBarType={getMessageBarType(this.state.notification.type)}
+                        styles={{ root: { width: 400 } }}
+                        onDismiss={() => this.setState({ notification: undefined })}
+                      >
+                        {getTranslation(this.state.notification.message)}
+                      </MessageBar>
+                    )}
+                    <SmartUiComponent
+                      disabled={this.state.refreshResult?.isUpdateInProgress}
+                      descriptor={this.state.root as SmartUiDescriptor}
+                      currentValues={this.state.currentValues}
+                      onInputChange={this.onInputChange}
+                      onError={this.onError}
+                      getTranslation={getTranslation}
```

Comment:
This might work for self-serve, but seems a bit too constraining for DE: does this mean for nested components we have to punch holes through all the intermediate ones? Ideally, we would want to just import the translation function from the "i18n" package anywhere in our code and use it like a global thing accessible from anywhere. Or have the top `<Translation>` node by a react Provider.

Reply:
will tackle this in separate PR.